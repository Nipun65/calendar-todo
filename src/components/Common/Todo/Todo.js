import { useEffect, useState } from 'react';
import Button from '../../UI/Button/Button';
import Checkbox from '../../UI/Checkbox/Chekbox';
import DialogBox from '../../UI/Modal/Modal';
import styles from './Todo.module.css';

function Todo({ date, Month }) {
  let gettodos = JSON.parse(window.localStorage.getItem('todos'));
  if (!gettodos) {
    gettodos = {};
  }
  const [todoObj, setTodoObj] = useState(gettodos);
  const [touched, setTouched] = useState(false);
  const [isValid, setValid] = useState(true);
  const [openModal, setModal] = useState(false);
  const [indexToEdit, setIndex] = useState(0);

  const todayDate = `${date.selectedDate.userDate}-${
    Month[date.selectedDate.userMonth]
  }-${date.selectedDate.userYear}`;

  useEffect(() => {
    setValid(true);
  }, [date]);

  const handleChange = () => {
    if (document.getElementById('todoText').value.length) {
      setValid(true);
    } else {
      setValid(false);
    }
  };

  const tempObj = todoObj;
  const addTodo = () => {
    setTouched(true);
    if (document.getElementById('todoText').value.length === 0) {
      setValid(false);
    } else {
      if (tempObj[todayDate] === undefined) {
        tempObj[todayDate] = [];
      }
      tempObj[todayDate].push({
        task: document.getElementById('todoText').value,
        isCompleted: false,
      });
      document.getElementById('todoText').value = '';
      window.localStorage.setItem('todos', JSON.stringify(tempObj));
      setTodoObj({ ...tempObj });
      setValid(true);
    }
  };

  const handleDelete = (index) => {
    todoObj[todayDate].splice(index, 1);
    window.localStorage.setItem('todos', JSON.stringify(tempObj));
    setTodoObj({ ...todoObj });
  };

  const handleComplete = (index) => {
    todoObj[todayDate][index].isCompleted =
      !todoObj[todayDate][index].isCompleted;
    window.localStorage.setItem('todos', JSON.stringify(tempObj));
    setTodoObj({ ...todoObj });
  };

  const handleUpdate = (index) => {
    setModal(true);
    setIndex(index);
  };

  const handleEditedValue = (value) => {
    todoObj[todayDate][indexToEdit].task = value;
    window.localStorage.setItem('todos', JSON.stringify(tempObj));
    setTodoObj({ ...todoObj });
  };

  return (
    <>
      <div className="has-text-centered pt-4 is-size-4-mobile is-size-3-tablet is-size-2-desktop has-text-weight-bold">
        To-do List
      </div>
      <div className="has-text-centered p-6">
        <div className={`control is-centered container ${styles.inputdiv}`}>
          <input
            id="todoText"
            className="input"
            type="text"
            placeholder="Enter Todo"
            onFocus={() => {
              setTouched(false);
            }}
            onChange={handleChange}
          />
          {!isValid && touched && (
            <p className="help is-danger has-text-weight-bold is-pulled-left px-1">
              This field is required
            </p>
          )}
        </div>

        <div className="pt-6">
          <Button
            classes={`button is-success ${styles.addtodobtn}`}
            onClick={addTodo}
            textContent="Add Todo"
          />
        </div>
      </div>
      <div className={styles['main-div']}>
        {todoObj[todayDate] &&
          todoObj[todayDate].map((dataValue, index) => (
            <div
              key={Math.random()}
              className={`p-4 m-2 ${styles['todo-div']}`}
            >
              <Checkbox
                onClick={() => handleComplete(index)}
                classes="is-inline"
                checked={dataValue.isCompleted}
              />

              <li
                className={`pl-3 ${styles.todocontent} ${styles.liele}`}
                style={
                  dataValue.isCompleted
                    ? { textDecoration: 'line-through' }
                    : {}
                }
              >
                {dataValue.task}
              </li>

              <div className="is-pulled-right">
                <Button
                  onClick={() => handleUpdate(index)}
                  classes={`fas fa-edit ${styles.editbutton}`}
                />

                <Button
                  onClick={() => handleDelete(index)}
                  classes="delete button is-danger"
                />
              </div>
            </div>
          ))}
      </div>
      {openModal && (
        <DialogBox
          open={openModal}
          setModal={setModal}
          editValue={handleEditedValue}
          currentValue={todoObj[todayDate][indexToEdit].task}
          title="Update Todo"
        />
      )}

      {(!todoObj[todayDate] || !todoObj[todayDate].length) && (
        <div className="has-text-centered subtitle is-4  has-text-weight-bold">
          No Todos Yet
        </div>
      )}
    </>
  );
}

export default Todo;
