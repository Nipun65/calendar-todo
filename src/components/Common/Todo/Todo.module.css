.linethrough {
  text-decoration: line-through;
}

.input-div {
  display: flex;
  flex-direction: column;
}

.background-color {
  background-color: lightgray;
  width: 50%;
  margin: 20px auto;
  word-break: break-all;
}

.editbutton {
  margin: -8px 15px 0px 0px;
}

.required {
  float: left;
}
