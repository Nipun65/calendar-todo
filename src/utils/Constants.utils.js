export const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export const DROPDOWNLIST = ['Month', 'Year'];

export const DAYS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

export const DATE = {
  year: 2023,
  month: 0,
  date: 31,
  currentMonth: 0,
  currentYear: 2023,
  selectedDate: { userYear: 2023, userMonth: 0, userDate: 31 },
};
