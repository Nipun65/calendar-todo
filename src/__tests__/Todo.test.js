import renderer from 'react-test-renderer';
import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Todo from '../components/Common/Todo/Todo';
import { MONTHS } from '../utils/Constants.utils';

const date = {
  year: 2023,
  month: 0,
  date: 31,
  currentMonth: 0,
  currentYear: 2023,
  selectedDate: { userYear: 2023, userMonth: 0, userDate: 31 },
};

afterEach(() => {
  cleanup();
});

it('should match the snapshot', () => {
  const tree = renderer.create(<Todo date={date} Month={MONTHS} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render button Component', () => {
  render(<Todo date={date} Month={MONTHS} />);
  const mainDiv = screen.getByTestId('todo');
  expect(mainDiv).toBeInTheDocument();
  expect(mainDiv).toHaveTextContent('To-do List');
  const notodotxt = screen.queryByText('No Todos Yet');
  expect(notodotxt).toBeInTheDocument();
});
