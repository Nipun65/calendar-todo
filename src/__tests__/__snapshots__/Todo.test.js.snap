// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
Array [
  <div
    className="has-text-centered pt-4 is-size-4-mobile is-size-3-tablet is-size-2-desktop has-text-weight-bold"
    data-testid="todo"
  >
    To-do List
  </div>,
  <div
    className="has-text-centered p-6"
  >
    <div
      className="control is-centered container inputdiv"
    >
      <input
        className="input"
        id="todoText"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter Todo"
        type="text"
      />
    </div>
    <div
      className="pt-6"
    >
      <button
        className="button is-success addtodobtn is-clickable"
        data-testid="button"
        onClick={[Function]}
        type="button"
      >
        Add Todo
      </button>
    </div>
  </div>,
  <div
    className="main-div"
  />,
  <div
    className="has-text-centered subtitle is-4 has-text-weight-bold"
  >
    No Todos Yet
  </div>,
]
`;
